hivemind/utils/grpc.py

```diff
@@ -160,6 +161,46 @@ def top(self) -> ValueError:
         raise ValueError(f"Please use {self.__class__.__name__}.get_stub to get or create stubs")
 
 
+def quantile_encode_torch_approx(weight, n_bits: int):
+    n_bins = 2 ** n_bits
+    borders = torch.as_tensor(quantile_qq_approximation(weight.numpy(), n_bins + 1)[1:-1])
+    quant_weight = torch.clamp_(torch.bucketize(weight, borders), 0, n_bins - 1)
+    bin_sums = torch.zeros(n_bins).scatter_add_(0, quant_weight.flatten(), weight.flatten())
+    bin_counts = torch.clamp_min_(torch.bincount(quant_weight.flatten(), minlength=n_bins), 1)
+    lookup = bin_sums / bin_counts
+    return quant_weight, lookup
+
+
+def quantile_qq_approximation(array, n_quantiles, min_chunk_size: int = 10 ** 5, **kwargs):
+    """ Estimate uniform quantiles of data using quantile-of-quantiles. Runs in parallel. """
```

Comment:
Might be just my ignorance, but what are _uniform_ quantiles and how are they different from regular quantiles?

Reply:
this technique specifically refers to estimating n+1 quantiles in a uniform symmetric grid in [0, 1/n, 2/n, ..., n/n]